src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs

```diff
@@ -270,5 +271,123 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
+                default:
+                    return null;
+            }
+        }
+
+        /// <summary>
+        /// Gets the entity type of the <paramref name="segment"/>.
+        /// </summary>
+        /// <param name="segment">The target <see cref="ODataOperationSegment"/>.</param>
+        /// <returns>The entity type of the target <paramref name="segment"/>.</returns>
+        private static IEdmEntityType EntityTypeFromOperationSegment(this ODataSegment segment)
+        {
+            CheckArgumentNull(segment, nameof(segment));
+
+            if (segment is ODataOperationSegment operationSegment)
+            {
+                IEdmOperationParameter bindingParameter = operationSegment.Operation.Parameters.First();
+                IEdmTypeReference bindingType = bindingParameter.Type;
+
+                if (bindingType.IsCollection())
+                {
+                    bindingType = bindingType.AsCollection().ElementType();
+                }
+
+                return bindingType.AsEntity().EntityDefinition();
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Attempts to add the specified <paramref name="path"/> and <paramref name="pathItem"/> to the <paramref name="pathItems"/> dictionary. 
+        /// </summary>
+        /// <param name="pathItems">The target dictionary.</param>
+        /// <param name="context">The OData context</param>
+        /// <param name="path">The key to be added.</param>
+        /// <param name="pathItem">The value to be added.</param>
+        /// <returns>true when the key and/or value are successfully added/updated to the dictionary; 
+        /// false when the dictionary already contains the specified key, and nothing gets added.</returns>
+        internal static bool TryAddPath(this IDictionary<string, OpenApiPathItem> pathItems,
+            ODataContext context,
+            ODataPath path,
+            OpenApiPathItem pathItem)
+        {
+            CheckArgumentNull(pathItems, nameof(pathItems));
+            CheckArgumentNull(context, nameof(context));
+            CheckArgumentNull(path, nameof(path));
+            CheckArgumentNull(pathItem, nameof(pathItem));
+
+            OpenApiConvertSettings settings = context.Settings.Clone();
+            settings.EnableKeyAsSegment = context.KeyAsSegment;
+
+            string pathName = path.PathTemplate ?? path.GetPathItemName(settings);
+
+            if (!pathItems.TryAdd(pathName, pathItem))
+            {
+                if (path.LastSegment is not ODataOperationSegment lastSegment)
+                {
+                    Debug.WriteLine("Duplicate path: " + pathName);
+                    return false;
+                }
+
+                int secondLastSegmentIndex = 2;
+                if (path.Count < secondLastSegmentIndex)
+                {
+                    Debug.WriteLine($"Invalid path. Operation not bound to any entity. Path: {pathName}");
```

Comment:
we could have the same issue with an operation bound to a collection, and have a collection at the root. E.g. have one activate method on directory object collections, and and another activate on users collection. All of a sudden /users/activate becomes a problem.

Reply:
_/users/activate_ will not fall within that conditional statement as the no. of segments will be equal to 2